src/data/setoid.lean

```diff
@@ -0,0 +1,580 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Amelia Livingston, Bryan Gin-ge Chen
+-/
+
+import data.quot data.set.lattice data.fintype order.galois_connection
+
+/-!
+# Equivalence relations
+
+The first section of the file defines the complete lattice of equivalence relations 
+on a type, results about the inductively defined equivalence closure of a binary relation, 
+and the analogues of some isomorphism theorems for quotients of arbitrary types.
+
+The second section comprises properties of equivalence relations viewed as partitions, and 
+a function `bell` that computes Bell numbers - the numbers of possible partitions of a finite 
+set - along with some preliminary results about `bell`.
+
+## Implementation notes
+
+The function `rel` and lemmas ending in ' make it easier to talk about different 
+equivalence relations on the same type.
+
+The complete lattice instance for equivalence relations could have been defined by lifting 
+the Galois insertion of equivalence relations on α into binary relations on α, and then using
+`complete_lattice.copy` to define a complete lattice instance with more appropriate 
+definitional equalities (a similar example is `filter.lattice.complete_lattice` in 
+`order/filter/basic.lean`). This does not save space, however, and is less clear.
+
+Partitions are not defined as a separate structure here; users are encouraged to
+reason about them using the existing `setoid` and its infrastructure.
+
+Eventually there should be a proof here that Bell numbers are the number of possible
+partitions of a finite set.
+
+## References
+
+* https://en.wikipedia.org/wiki/Bell_triangle
+
+## Tags
+
+setoid, equivalence, iseqv, relation, equivalence relation, partition, equivalence 
+class, Bell number, Bell triangle
+-/
+variables {α : Type*} {β : Type*}
+
+open lattice 
+
+namespace setoid
+
+/-- A version of setoid.r that takes the equivalence relation as an explicit argument. -/
+def rel (r : setoid α) := @setoid.r _ r
+
+@[extensionality] lemma ext' {r s : setoid α} (H : ∀ a b, r.rel a b ↔ s.rel a b) : 
+  r = s := ext H
+
+lemma ext_iff {r s : setoid α} : r = s ↔ ∀ a b, r.rel a b ↔ s.rel a b :=
+⟨λ h a b, h ▸ iff.rfl, ext'⟩
+
+/-- Defining '≤' for equivalence relations. -/
+instance : has_le (setoid α) := ⟨λ r s, ∀ x y, r.rel x y → s.rel x y⟩
+
+theorem le_def {r s : setoid α} : r ≤ s ↔ ∀ {x y}, r.rel x y → s.rel x y := iff.rfl
+
+@[refl] lemma refl' (r : setoid α) (x) : r.rel x x := r.2.1 x
+@[symm] lemma symm' (r : setoid α) : ∀ {x y}, r.rel x y → r.rel y x := λ _ _ h, r.2.2.1 h
+@[trans] lemma trans' (r : setoid α) : ∀ {x y z}, r.rel x y → r.rel y z → r.rel x z := 
+λ  _ _ _ hx hy, r.2.2.2 hx hy
+
+/-- The kernel of a function is an equivalence relation. -/
+def ker (f : α → β) : setoid α :=
+⟨λ x y, f x = f y, ⟨λ x, rfl, λ _ _ h, h.symm, λ _ _ _ h1 h2, h1.trans h2⟩⟩ 
+
+/-- The kernel of the quotient map induced by an equivalence relation r equals r. -/
```

Comment:
```suggestion
/-- The kernel of the quotient map induced by an equivalence relation on r equals r. -/
```

Reply:
Actually, I'm confused, are you sure? I meant the equiv reln is called r, to avoid saying 'equals the original equiv reln' or something

Reply:
Ah, I see what you mean. I agree, this should be reverted.

Reply:
Oh yes, I misread this.